File: IntroductoryProblems/gridPaths.py
s = input()


def createGrid(path):
    [x,y] = [0,0]
    grid = [[0 for _ in range(7)] for _ in range(7)]
    grid[x][y] = 1
    for c in path:
        if c=='D':
            x+=1
        elif c=='U':
            x-=1
        elif c=='R':
            y+=1
        elif c=='L':
            y-=1
        if x<0 or x>6 or y<0 or y>6 or grid[x][y]==1:
            return None, None
        else:
            grid[x][y]=1
    return grid,[x,y]

def nextPossibilities(grid, position):
    [x,y]=position
    ps, nn = checkNext(grid,position)
    must = [0 for _ in range(len(ps))]
    imps = [0 for _ in range(len(ps))]

    for i in range(len(ps)):
        num_pss, _ = checkNext(grid,nn[i])
        num_ns = len(num_pss)
        if num_ns==0:
            return None
        elif nn[i]==[6,0]:
            imps[i]=1
        elif num_ns==1:
            must[i]=1
    if sum(must)>1:
        return None
    elif sum(must)==1:
        if sum(imps)==0 or must.index(1)!=imps.index(1):
            ps = [ps[must.index(1)]]
    elif sum(imps)!=0:
#        suma =0
#        for i  in range(7):
#            suma+=sum(grid[i])
#        if suma!=47:
        for i in reversed(range(len(imps))):
            if imps[i]==1:
                ps.remove(ps[i])

    return ps

def checkNext(grid, position):
    [x,y]=position
    ps = ['D', 'U', 'R', 'L']
    nn = []
    res = []
    t = 0
    for i in [x+1,x-1]:
        if i>=0 and i<7 and grid[i][y]==0:
            res.append(ps[t])
            nn.append([i,y])
        t+=1

    for j in [y+1,y-1]:
        if j>=0 and j<7 and grid[x][j]==0:
            res.append(ps[t])
            nn.append([x,j])
        t += 1
    return res,nn

def nicePrint(list,n):
    t = 0
    f = open("/path.txt", 'a+')

    f.write('[')
    while t<len(list):
        text = ""
        for _ in range(n):
            if t==len(list):
                break
            text+=("'"+list[t]+"', ")
            t+=1
        f.write(text)
    f.write(']')
    f.close()

def findPath(s):
    next_paths = [""]
    for i in range(47):
        paths = next_paths

        print(i, len(paths))
        next_paths = []

        for path in paths:
            grid, position = createGrid(path)
            if grid and position:
                ps = nextPossibilities(grid, position)
            if ps:
                if s[i]=='?':
                    for p in ps:
                        next_paths.append(path+p)
                elif ps.count(s[i])==1:
                    next_paths.append(path+s[i])

    paths = next_paths
    next_paths = []

    for path in paths:
        ps = ['D','U','R','L']
        number = [path.count(p) for p in ps]

        if number[0]-number[1]!=6 and number[2]-number[3]!=0:
            continue
        elif number[0]-number[1]==5 and (s[47]=='?' or s[47]=='D'):
            next_paths.append(path+'D')
        elif number[2]-number[3]==1 and (s[47]=='?' or s[47]=='L'):
            next_paths.append(path+'L')

    print(len(next_paths))

findPath(s)

File: IntroductoryProblems/gridPaths.cpp

#include <list>
#include <stdlib.h>
#include <iostream>

struct Solution {
    int grid[7][7];
    char path[48]="???????????????????????????????????????????????";
    int suma = 0;
    int x = 0;
    int y = 0;
};

bool inRange(int n)
{
    bool res = (0<=n && n<7);
    return res;
}

int * checkNeightbors(int grid[7][7], int x, int y, int res[4])
{
    if (inRange(x-1) && grid[x-1][y]==0) {
        res[0] = 1;
    } if (inRange(x+1) && grid[x+1][y]==0) {
        res[1] = 1;
    } if (inRange(y-1) && grid[x][y-1]==0) {
        res[2] = 1;
    } if (inRange(y+1) && grid[x][y+1]==0) {
        res[3] = 1;
    }
    return res;
}

int numNeightbors(int grid[7][7], int x, int y)
{
    int res = 0;
    if (inRange(x-1) && grid[x-1][y]==0) {
        res++;
    } if (inRange(x+1) && grid[x+1][y]==0) {
        res++;
    } if (inRange(y-1) && grid[x][y-1]==0) {
        res++;
    } if (inRange(y+1) && grid[x][y+1]==0) {
        res++;
    }
    return res;
}

int sum(int list[4])
{
    int res = 0;
    for (int i=0;i<4;i++) {
        if (list[i]==1)
            res++;
    }
    return res;
}

bool compList(int must[4], int imps[4])
{
    for (int i=0; i<4;i++) {
        if (imps[i]==1 && must[i]==1)
            return true;
    }
    return false;
}

char to_string(int n)
{
    if (n==1) {
        return '1';
    } else {
        return '0';
    }
}

void printGrid(int grid[7][7])
{
    char pr[7];
    for (int i=0; i<7;i++) {
        for (int j = 0; j<7; j++) {
            pr[j]=to_string(grid[i][j]);
        }
        std::cout << " " << pr << std::endl;
    }
    std::cout << "\n";
}

void tryNextMoves(Solution &s, char input[48], int respuesta[48])
{
    respuesta[s.suma]++;
    if (s.suma==47) {
        if (input[47]=='?') {
            respuesta[48]++;
        }
        else if (input[47]=='L' && s.x==6 && s.y==1) {
            respuesta[48]++;
        }
        else if (input[47]=='D' && s.x==5 && s.y==0) {
            respuesta[48]++;
        }
    }

    int res[4] = {0,0,0,0};
    int *nn = checkNeightbors(s.grid,s.x,s.y, res);
    int must[4] = {0,0,0,0};
    int imps[4] = {0,0,0,0};


    if (nn[0]==1) {
        int n = numNeightbors(s.grid,s.x-1,s.y);
        if (n==0) {
            return;
        } else if (n==1) {
            must[0]=1;
        }
    }
    if (nn[1]==1) {
        int n = numNeightbors(s.grid,s.x+1,s.y);
        if (n==0) {
            return;
        } else if (s.x+1==6 && s.y==0) {
            imps[1]=1;
        } else if (n==1) {
            must[1]=1;
        }
    }
    if (nn[2]==1) {
        int n = numNeightbors(s.grid,s.x,s.y-1);
        if (n==0) {
            return;
        } else if (s.x==6 && s.y-1==0) {
            imps[2]=1;
        } else if (n==1) {
            must[2]=1;
        }
    }
    if (nn[3]==1) {
        int n = numNeightbors(s.grid,s.x,s.y+1);
        if (n==0) {
            return;
        } else if (n==1) {
            must[3]=1;
        }
    }

    if (sum(must)>1) {
        return;
    }
    else if (sum(must)==1) {
        if (sum(imps)==0 || !(compList(must,imps))) {
            nn = must;
        }
    }
    else if (imps[1]!=0) {
        nn[1]=0;
    }
    else if (imps[2]==1) {
        nn[2]=0;
    }

    if (input[s.suma]=='?') {
        if (nn[0]==1) {
            Solution temp = s;
            temp.path[temp.suma]='U';
            temp.suma++;
            temp.x--;
            temp.grid[temp.x][temp.y]=1;
            tryNextMoves(temp,input, respuesta);
        }
        if (nn[1]==1) {
            Solution temp = s;
            temp.path[temp.suma]='D';
            temp.suma++;
            temp.x++;
            temp.grid[temp.x][temp.y]=1;
            tryNextMoves(temp,input, respuesta);
        }
        if (nn[2]==1) {
            Solution temp = s;
            temp.path[temp.suma]='L';
            temp.suma++;
            temp.y--;
            temp.grid[temp.x][temp.y]=1;
            tryNextMoves(temp,input, respuesta);
        }
        if (nn[3]==1) {
            Solution temp = s;
            temp.path[temp.suma]='R';
            temp.suma++;
            temp.y++;
            temp.grid[temp.x][temp.y]=1;
            tryNextMoves(temp,input, respuesta);
        }
    }
    else if (input[s.suma]=='U') {
        if (nn[0]==1) {
            Solution temp = s;
            temp.path[temp.suma]='U';
            temp.suma++;
            temp.x--;
            temp.grid[temp.x][temp.y]=1;
            tryNextMoves(temp,input, respuesta);
        }
    }
    else if (input[s.suma]=='D') {
        if (nn[1]==1) {
            Solution temp = s;
            temp.path[temp.suma]='D';
            temp.suma++;
            temp.x++;
            temp.grid[temp.x][temp.y]=1;
            tryNextMoves(temp,input, respuesta);
        }
    }
    else if (input[s.suma]=='L') {
        if (nn[2]==1) {
            Solution temp = s;
            temp.path[temp.suma]='L';
            temp.suma++;
            temp.y--;
            temp.grid[temp.x][temp.y]=1;
            tryNextMoves(temp,input, respuesta);
        }
    }
    else if (input[s.suma]=='R') {
        if (nn[3]==1) {
            Solution temp = s;
            temp.path[temp.suma]='R';
            temp.suma++;
            temp.y++;
            temp.grid[temp.x][temp.y]=1;
            tryNextMoves(temp,input, respuesta);
        }
    }

}

int main()
{

    int respuesta[49];
    for (int i =0;i<49;i++) {
        respuesta[i]=0;
    }

    char input[48];
    std::cin >> input;
    //std::cout <<input[47] << std::endl;

    Solution s;
    for (int i = 0; i<7; i++) {
        for (int j = 0; j<7; j++) {
            s.grid[i][j]=0;
        }
    }
    s.grid[0][0]=1;

    tryNextMoves(s,input,respuesta);

    std::cout << respuesta[48] << std::endl;




}